inject/hk2/src/main/java/org/glassfish/jersey/inject/hk2/SupplierFactoryBridge.java

```diff
@@ -85,9 +87,11 @@ public T provide() {
     @Override
     public void dispose(T instance) {
         if (disposable) {
-            DisposableSupplier<T> disposableSupplier = disposableSuppliers.get(instance);
-            disposableSupplier.dispose(instance);
-            disposableSuppliers.remove(instance);
+            synchronized (disposableSuppliers) {
+                DisposableSupplier<T> disposableSupplier = disposableSuppliers.get(instance);
+                disposableSupplier.dispose(instance);
+                disposableSuppliers.remove(instance);
```

Comment:
This block is too wide - it synchronizes also the disposal of the supplier
Change this block:
```
DisposableSupplier<T> disposableSupplier = disposableSuppliers.remove(instance);
disposableSupplier.dispose(instance);
```

Reply:
I would still replace the current implementation with the one I proposed.
I would rather call a single method (`remove`) than two methods (`get` and `remove`) on a synchronized object.